wasmtime fails to compile with could not find `fde` in `isa` within: crates/environ/src/data_structures.rs:18:41 on arm

there's a recent regression in the code: 

```
     Running `rustc --crate-name wasmtime_environ --edition=2018 crates/environ/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts --crate-type lib --emit=dep-info,metadata,link -C debuginfo=2 -C metadata=3f3bb9806de25bbc -C extra-filename=-3f3bb9806de25bbc --out-dir /tmp/wasmtime/target/armv7-unknown-linux-gnueabihf/debug/deps --target armv7-unknown-linux-gnueabihf -C linker=armv7a-unknown-linux-gnueabihf-gcc -C incremental=/tmp/wasmtime/target/armv7-unknown-linux-gnueabihf/debug/incremental -L dependency=/tmp/wasmtime/target/armv7-unknown-linux-gnueabihf/debug/deps -L dependency=/tmp/wasmtime/target/debug/deps --extern anyhow=/tmp/wasmtime/target/armv7-unknown-linux-gnueabihf/debug/deps/libanyhow-a7d5a95d013f511f.rmeta --extern base64=/tmp/wasmtime/target/armv7-unknown-linux-gnueabihf/debug/deps/libbase64-1fc3337a027c3d46.rmeta --extern bincode=/tmp/wasmtime/target/armv7-unknown-linux-gnueabihf/debug/deps/libbincode-0cfacf3864ed393b.rmeta --extern cranelift_codegen=/tmp/wasmtime/target/armv7-unknown-linux-gnueabihf/debug/deps/libcranelift_codegen-fe4d1e7b8c528231.rmeta --extern cranelift_entity=/tmp/wasmtime/target/armv7-unknown-linux-gnueabihf/debug/deps/libcranelift_entity-6f3c3d46881e39ec.rmeta --extern cranelift_wasm=/tmp/wasmtime/target/armv7-unknown-linux-gnueabihf/debug/deps/libcranelift_wasm-c3fa487d86de75b7.rmeta --extern directories=/tmp/wasmtime/target/armv7-unknown-linux-gnueabihf/debug/deps/libdirectories-32547e134205d970.rmeta --extern errno=/tmp/wasmtime/target/armv7-unknown-linux-gnueabihf/debug/deps/liberrno-aa6b44df9b99329e.rmeta --extern file_per_thread_logger=/tmp/wasmtime/target/armv7-unknown-linux-gnueabihf/debug/deps/libfile_per_thread_logger-c9cdc123d34b0b2e.rmeta --extern indexmap=/tmp/wasmtime/target/armv7-unknown-linux-gnueabihf/debug/deps/libindexmap-062ef480b02be0ea.rmeta --extern libc=/tmp/wasmtime/target/armv7-unknown-linux-gnueabihf/debug/deps/liblibc-45ed8421ceb1d133.rmeta --extern log=/tmp/wasmtime/target/armv7-unknown-linux-gnueabihf/debug/deps/liblog-4ab7a3e889d23791.rmeta --extern more_asserts=/tmp/wasmtime/target/armv7-unknown-linux-gnueabihf/debug/deps/libmore_asserts-4602433f0d9089cb.rmeta --extern rayon=/tmp/wasmtime/target/armv7-unknown-linux-gnueabihf/debug/deps/librayon-8631c1a88fa6e275.rmeta --extern serde=/tmp/wasmtime/target/armv7-unknown-linux-gnueabihf/debug/deps/libserde-198981066e594298.rmeta --extern sha2=/tmp/wasmtime/target/armv7-unknown-linux-gnueabihf/debug/deps/libsha2-39588d910b40a085.rmeta --extern thiserror=/tmp/wasmtime/target/armv7-unknown-linux-gnueabihf/debug/deps/libthiserror-8289a6a6c6a57ad6.rmeta --extern toml=/tmp/wasmtime/target/armv7-unknown-linux-gnueabihf/debug/deps/libtoml-6acf23745dbe4062.rmeta --extern wasmparser=/tmp/wasmtime/target/armv7-unknown-linux-gnueabihf/debug/deps/libwasmparser-6bba5d41f47f7b65.rmeta --extern zstd=/tmp/wasmtime/target/armv7-unknown-linux-gnueabihf/debug/deps/libzstd-4a14a693b151a604.rmeta -L native=/tmp/wasmtime/target/armv7-unknown-linux-gnueabihf/debug/build/zstd-sys-8fe68259522e1833/out`
error[E0432]: unresolved import `cranelift_codegen::isa::fde`
  --> crates/environ/src/data_structures.rs:18:41
   |
18 |         pub use cranelift_codegen::isa::fde::map_reg;
   |                                         ^^^ could not find `fde` in `isa`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0432`.
error: could not compile `wasmtime-environ`.

```

caused by: https://github.com/bytecodealliance/wasmtime/commit/d3df27500334625f463ea0f8134ea28395a5e6a0

reverting the commit locally fixes the issue for me.

--- a/cranelift/codegen/src/isa/fde.rs
+++ b/cranelift/codegen/src/isa/fde.rs
@@ -1,0 +1,14 @@
+//! Support for FDE data generation.
+use thiserror::Error;
+
+/// Enumerate the errors possible in mapping Cranelift registers to their DWARF equivalent.
+#[allow(missing_docs)]
+#[derive(Error, Debug)]
+pub enum RegisterMappingError {
+    #[error("unable to find bank for register info")]
+    MissingBank,
+    #[error("register mapping is currently only implemented for x86_64")]
+    UnsupportedArchitecture,
+    #[error("unsupported register bank: {0}")]
+    UnsupportedRegisterBank(&'static str),
+}

--- a/cranelift/codegen/src/isa/mod.rs
+++ b/cranelift/codegen/src/isa/mod.rs
@@ -56,6 +56,8 @@ use crate::binemit;
 use crate::flowgraph;
 use crate::ir;
 use crate::isa::enc_tables::Encodings;
+#[cfg(feature = "unwind")]
+use crate::isa::fde::RegisterMappingError;
 use crate::regalloc;
 use crate::result::CodegenResult;
 use crate::settings;
@@ -74,11 +76,8 @@ mod riscv;
 #[cfg(feature = "x86")]
 mod x86;
 
-#[cfg(all(feature = "x86", feature = "unwind"))]
-/// Expose the register-mapping functionality necessary for exception handling, debug, etc.
-pub mod fde {
-    pub use super::x86::map_reg;
-}
+#[cfg(feature = "unwind")]
+pub mod fde;
 
 #[cfg(feature = "arm32")]
 mod arm32;
@@ -259,6 +258,12 @@ pub trait TargetIsa: fmt::Display + Send + Sync {
 
     /// Get a data structure describing the registers in this ISA.
     fn register_info(&self) -> RegInfo;
+
+    #[cfg(feature = "unwind")]
+    /// Map a Cranelift register to its corresponding DWARF register.
+    fn map_dwarf_register(&self, _: RegUnit) -> Result<u16, RegisterMappingError> {
+        Err(RegisterMappingError::UnsupportedArchitecture)
+    }
 
     /// Returns an iterator over legal encodings for the instruction.
     fn legal_encodings<'a>(

--- a/cranelift/codegen/src/isa/x86/fde.rs
+++ b/cranelift/codegen/src/isa/x86/fde.rs
@@ -2,6 +2,7 @@
 
 use crate::binemit::{FrameUnwindOffset, FrameUnwindSink, Reloc};
 use crate::ir::{FrameLayoutChange, Function};
+use crate::isa::fde::RegisterMappingError;
 use crate::isa::{CallConv, RegUnit, TargetIsa};
 use alloc::vec::Vec;
 use core::convert::TryInto;
@@ -10,7 +11,6 @@ use gimli::write::{
     FrameDescriptionEntry, FrameTable, Result, Writer,
 };
 use gimli::{Encoding, Format, LittleEndian, Register, X86_64};
-use thiserror::Error;
 
 pub type FDERelocEntry = (FrameUnwindOffset, Reloc);
 
@@ -135,16 +135,6 @@ pub fn map_reg(
         "FloatRegs" => Ok(X86_XMM_REG_MAP[(reg - bank.first_unit) as usize]),
         _ => Err(RegisterMappingError::UnsupportedRegisterBank(bank.name)),
     }
-}
-
-#[derive(Error, Debug)]
-pub enum RegisterMappingError {
-    #[error("unable to find bank for register info")]
-    MissingBank,
-    #[error("register mapping is currently only implemented for x86_64")]
-    UnsupportedArchitecture,
-    #[error("unsupported register bank: {0}")]
-    UnsupportedRegisterBank(&'static str),
 }
 
 fn to_cfi(

--- a/cranelift/codegen/src/isa/x86/mod.rs
+++ b/cranelift/codegen/src/isa/x86/mod.rs
@@ -22,6 +22,8 @@ use crate::binemit::{FrameUnwindKind, FrameUnwindSink};
 use crate::ir;
 use crate::isa::enc_tables::{self as shared_enc_tables, lookup_enclist, Encodings};
 use crate::isa::Builder as IsaBuilder;
+#[cfg(feature = "unwind")]
+use crate::isa::{fde::RegisterMappingError, RegUnit};
 use crate::isa::{EncInfo, RegClass, RegInfo, TargetIsa};
 use crate::regalloc;
 use crate::result::CodegenResult;
@@ -89,6 +91,11 @@ impl TargetIsa for Isa {
 
     fn register_info(&self) -> RegInfo {
         registers::INFO.clone()
+    }
+
+    #[cfg(feature = "unwind")]
+    fn map_dwarf_register(&self, reg: RegUnit) -> Result<u16, RegisterMappingError> {
+        map_reg(self, reg).map(|r| r.0)
     }
 
     fn encoding_info(&self) -> EncInfo {

--- a/crates/debug/src/frame.rs
+++ b/crates/debug/src/frame.rs
@@ -1,6 +1,5 @@
 use std::collections::HashMap;
 use wasmtime_environ::entity::EntityRef;
-use wasmtime_environ::isa::fde::map_reg;
 use wasmtime_environ::isa::{CallConv, TargetIsa};
 use wasmtime_environ::wasm::DefinedFuncIndex;
 use wasmtime_environ::{FrameLayoutChange, FrameLayouts};
@@ -19,8 +18,8 @@ fn to_cfi(
 ) -> Option<CallFrameInstruction> {
     Some(match change {
         FrameLayoutChange::CallFrameAddressAt { reg, offset } => {
-            let mapped = match map_reg(isa, *reg) {
-                Ok(r) => r,
+            let mapped = match isa.map_dwarf_register(*reg) {
+                Ok(r) => Register(r),
                 Err(_) => return None,
             };
             let offset = (*offset) as i32;
@@ -41,8 +40,8 @@ fn to_cfi(
         FrameLayoutChange::RegAt { reg, cfa_offset } => {
             assert!(cfa_offset % -8 == 0);
             let cfa_offset = *cfa_offset as i32;
-            let mapped = match map_reg(isa, *reg) {
-                Ok(r) => r,
+            let mapped = match isa.map_dwarf_register(*reg) {
+                Ok(r) => Register(r),
                 Err(_) => return None,
             };
             CallFrameInstruction::Offset(mapped, cfa_offset)

--- a/crates/debug/src/transform/expression.rs
+++ b/crates/debug/src/transform/expression.rs
@@ -5,7 +5,6 @@ use more_asserts::{assert_le, assert_lt};
 use std::collections::{HashMap, HashSet};
 use wasmtime_environ::entity::EntityRef;
 use wasmtime_environ::ir::{StackSlots, ValueLabel, ValueLabelsRanges, ValueLoc};
-use wasmtime_environ::isa::fde::map_reg;
 use wasmtime_environ::isa::TargetIsa;
 use wasmtime_environ::wasm::{get_vmctx_value_label, DefinedFuncIndex};
 use wasmtime_environ::ModuleMemoryOffset;
@@ -79,7 +78,7 @@ fn translate_loc(
     use gimli::write::Writer;
     Ok(match loc {
         ValueLoc::Reg(reg) => {
-            let machine_reg = map_reg(isa, reg)?.0 as u8;
+            let machine_reg = isa.map_dwarf_register(reg)? as u8;
             Some(if machine_reg < 32 {
                 vec![gimli::constants::DW_OP_reg0.0 + machine_reg]
             } else {
@@ -120,8 +119,8 @@ fn append_memory_deref(
     // FIXME for imported memory
     match vmctx_loc {
         ValueLoc::Reg(vmctx_reg) => {
-            let reg = map_reg(isa, vmctx_reg)?;
-            writer.write_u8(gimli::constants::DW_OP_breg0.0 + reg.0 as u8)?;
+            let reg = isa.map_dwarf_register(vmctx_reg)? as u8;
+            writer.write_u8(gimli::constants::DW_OP_breg0.0 + reg)?;
             let memory_offset = match frame_info.vmctx_memory_offset() {
                 Some(offset) => offset,
                 None => {

--- a/crates/environ/src/data_structures.rs
+++ b/crates/environ/src/data_structures.rs
@@ -14,9 +14,6 @@ pub mod settings {
 
 pub mod isa {
     pub use cranelift_codegen::isa::{CallConv, RegUnit, TargetFrontendConfig, TargetIsa};
-    pub mod fde {
-        pub use cranelift_codegen::isa::fde::map_reg;
-    }
 }
 
 pub mod entity {

